### ISSUE TITLE
shipping with windows 10 21h2 or windows 11

### ISSUE THREAD

**Issue by ahmed-abdelrazek** [microsoft/wslg#347]
is this going to ship with windows 10 update 21h2 or windows 11 or both?!

**Comment by Biswa96** [comment]
This topic is being discussed here https://github.com/microsoft/wslg/discussions/82

**Comment by ahmed-abdelrazek** [comment]
@Biswa96 doesn't really have a concrete answer to my question 
will they ship it to windows 10 or not? i can't upgrade to windows 11

**Comment by spronovo** [comment]
WSLg will ships with Windows 11. At this time there are no plan to enable WSLg in Windows 10.

**Comment by ahmed-abdelrazek** [comment]
you developed and tested the feature for months on windows 10 and now telling us it's not coming just because little disappointed here  @spronovo

**Comment by crramirez** [comment]
> WSLg will ships with Windows 11. At this time there are no plan to enable WSLg in Windows 10.

Thank you, @spronovo, for your sincerity. Knowing this now helps us who develop WSL solutions for other users to be prepared and offer alternatives to ones that cannot upgrade to Windows 11, also ones (enterprise users) that won't wish to upgrade yet.

**Comment by exitdown** [comment]
yep, no suprises from microsoft then. always suspected the deliberately cagey language of "the next major release of windows"  very dissapointing still.

**Comment by thany** [comment]
> WSLg will ships with Windows 11. At this time there are no plan to enable WSLg in Windows 10.

That is essentially limiting this feature to the fortunate few who have been selected by Microsoft to be allowed to run Windows 11, for a feature that emulates scenarios that happily run on computers that are essentially decades old.

Do you not see the irony?

**Comment by Masamune3210** [comment]
I think you misunderstand how much work is involved with getting this running, it's not as simple as just making a plan

**Comment by ahmed-abdelrazek** [comment]
i think it is they didn't even bother to change the windows 10 screenshot in the readme file 
the feature was developed, tested and advertised under windows 10 all of the sudden after windows 11 announcement windows 10 can't run it i call bullshit https://devblogs.microsoft.com/commandline/the-initial-preview-of-gui-app-support-is-now-available-for-the-windows-subsystem-for-linux-2/

**Comment by exitdown** [comment]
upvote my 'enhancement request' and tell your freinds :) https://github.com/microsoft/wslg/issues/414

**Comment by Masamune3210** [comment]
If its such a important feature, why not just upgrade to 11 when it comes out? They already have said that you will be able to manually install 11 even if you don't meet the new requirements for upgrade

**Comment by ahmed-abdelrazek** [comment]
you wouldn't get drivers or system updates and your old drivers it's guarantee to work also you still need tpm 1.2 which i don't have and i don't like what windows 11 has to offer so far https://www.bleepingcomputer.com/news/microsoft/windows-11-may-not-get-security-updates-on-unsupported-devices/

**Comment by Mempler** [comment]
> If its such a important feature, why not just upgrade to 11 when it comes out? They already have said that you will be able to manually install 11 even if you don't meet the new requirements for upgrade

Straight up, not everyone has support for Win11, E.G I don't have an TPM module and my CPU `fx 8350` is simply not supported. I simply do not see a reason why i would ever want to have a TPM module implemented. the last time i've had ransomware was like... Never, or a virus, also something i haven't had in like 8 years. Only kids who crack random ass software get this shit or people who don't know what they're doing. but quite literally a useless module for technically served people.

Also, i cannot afford to spend 800 bucks for a new freaking computer.

**Comment by exitdown** [comment]
Im Sure the engineering team would be more than happy to allow the feature in a windows 10 feature update, and all they need to do that is enough people to request the feature. theres no need to do it in a rude way, just upvote the feature request. if we got enough upvotes they'd be able to justify the dev time. its that simple. management obvioulsy doesnt think there is enough people who care about it to warrant maintaining the feature on two codepaths, and if im being honest, comparatively speaking its a niche edgecase feature  already - but the target audience of this feature is most likley to also want to use it on windows 10. so obviously the solution is simply upvote the enhanceent request till theres enough justification..... no need to argue against or for. just upvote the enhancement and give them some raw numebrs. management want raw numbers.

**Comment by ahmed-abdelrazek** [comment]
do you know how many feature requests got thousands of upvotes on feedback hub and uservoice were there for years and no one cared 
upvotes don't mean much for example 
![image](https://user-images.githubusercontent.com/7858946/132988141-16474c90-40e6-4751-b5ce-7d66400fa74c.png)

also you seem to get that one wrong we are not talking about a feature they announced for windows 11 and we demanding to be on windows 10 too like wsa or something we are talking about a feature that was already planned developed and tested on windows 10 and they decided to make it exclusive to windows 11 it's literally on windows 10 insider builds right now so

**Comment by exitdown** [comment]
no- i was here from the beginning of wsl, an active tester, and wgl too - im simply stating that if you want the team to have any chance whatsoever of having their governance give them the authority to spend capex and opex resources on implementing it there has to be an obvious justification for that - upvoting an enhancement request is a way for the community to contribute to that justification.  - also adding tabs to file explorer is a core change to a core feature - this is an enhancement to a brand new feature thats still in active dev and not at "production quality" standard yet - big difference.

**Comment by ahmed-abdelrazek** [comment]
"spend capex and opex resources on implementing it there" that's my whole point it's already there

and my other point was no matter how many upvotes a feature gets if it aliened with microsoft vision they will do it no matter what and if not then tough luck

your point was keep upvoting and they will do it eventually then it became unless it's a core feature that needs a core change then upvotes don't matter
pick a lane

**Comment by thany** [comment]
@Masamune3210 

> I think you misunderstand how much work is involved with getting this running, it's not as simple as just making a plan

Can you explain why it is *so much more work* to get it to work on Windows 10, versus Windows 11?

**Comment by sabbyX** [comment]
> the feature was developed, tested and advertised under windows 10 all of the sudden after windows 11 announcement windows 10 can't run it.

@ahmed-abdelrazek you are likely missing out the fact that the build 21364 was part of Windows 11 development, technically they were testing WSLg under Windows 11, not Windows 10.

**Comment by ahmed-abdelrazek** [comment]
@sabbyX maybe you are right but i'm not the one that kept calling it windows 10 preview even after that
![image](https://user-images.githubusercontent.com/7858946/134339120-55a79c18-a40e-4f6c-90b2-9dea6c3bb7fa.png)
windows 11 started at 22000 
![image](https://user-images.githubusercontent.com/7858946/134339639-302ed4b2-61e3-496a-bed5-ecc89b434557.png)
yet
![image](https://user-images.githubusercontent.com/7858946/134340208-9edc8539-aab7-4668-9869-8abbf82e2791.png)
but
![image](https://user-images.githubusercontent.com/7858946/134340481-70d8b020-d65f-4038-8b96-99475c224d70.png)
which also required a 20000 build
![image](https://user-images.githubusercontent.com/7858946/134340645-b59cf8f5-805f-440d-9d3c-8baf2780c619.png)

and that's part of the problem i'm talking about no clear communication from Microsoft i guess everything is going to be more clear in a few weeks now after the official release for both windows 10 21h2 and windows 11

**Comment by exitdown** [comment]
i love how theres 73 downvotes on the fact they arent releasing for windows 10, but only 45 upvotes on my enhancement request. and people who presumably want the feature are instead attacking everyone arguing for simply requesting that it be a feature and lending weight to it with bullshit arguments about "it wont make any difference" and <insert straw man here> honestly its almost as if none of you have had any experience developing software in a commercial environment. its microsoft, they have corporate agenda and corporate governance. its that simple. if you give them some statisticis they can paint on a pretty graph to show the marketing douche who has been placed in a position to make decisions about the engineering of the products, then he might be inclined to authorise the work. if you dont do that, theres no way he will, because marketing douches dont listen to engineers, they listen to sales douches, and pie charts. its really simple stuff, and you guys bitch fighting about it internally are just an embarrassing cliche at this point. i give up.

**Comment by crramirez** [comment]
Enterprise adoption of WSL can make things change. In the past, they have backported, for example, WSL2 and other minor updates due to enterprise customers tend to be slower in upgrading their Windows.

They first need the numbers on how Windows 11 is being adopted. By the end of 2022, they'll know if many companies stay on Windows 10 until 2025 that use WSL and want WSLg. But if the adoption is very high, no backport will take place.

Until then, @exitdown, don't give up; these votes up will be relevant then. Not now.

**Comment by thany** [comment]
Well if Microsoft can make Windows 11 run on my machine, all will be forgiven (from my end).

**Comment by ahmed-abdelrazek** [comment]
just as a reminder https://www.windowscentral.com/microsoft-directstorage-coming-windows-11-and-windows-10

**Comment by boarder2** [comment]
> WSLg will ships with Windows 11. At this time there are no plan to enable WSLg in Windows 10.

This feels like such a huge missed target. WSL is there for developers and enterprise users. The majority of who won't be running Win 11 for years yet, especially given the hardware requirements for Win 11.

The early announcements and demos of WSLg were done on Win 10.  I feel like this goes against everything Microsoft has been doing to be more open and available everywhere.  What a complete let down.

**Comment by Biswa96** [comment]
There is something https://github.com/MicrosoftDocs/WSL/commit/a49448706f7c4569c72deea4d0874755339f8bc0

**Comment by jackblk** [comment]
> There is something [MicrosoftDocs/WSL@a494487](https://github.com/MicrosoftDocs/WSL/commit/a49448706f7c4569c72deea4d0874755339f8bc0)

Seems like it's bundled with WSL now, but the app itself requires build 22000 or above

![image](https://user-images.githubusercontent.com/12843654/136894162-69554ea6-aef4-41ee-b5a9-df32a8d31f1b.png)

**Comment by paulius-valiunas** [comment]
I might be able to upgrade, but the problem is I don't want to. Windows 11 has horrible UX (excessive padding, rounded corners wasting space everywhere, lack of taskbar customization, menus that move around and block any form of muscle memory etc.) and it will take a while for all these problems to be fixed, if they're ever fixed at all. I'm still hoping for some 3rd party customizations to appear, fixing some of the glaring UX problems that Microsoft is clearly not going to fix. Isn't it ironic that I found out about WSLG by watching the VS2022 launch event where most of the demos were made by Microsoft developers using Windows 10 with customized taskbars? Microsoft, even your own developers don't use Windows 11, so how can you expect other developers to adopt it early?

Enough rant. My point is, WSLG not supporting Windows 10 is a huge problem and _not_ only in corporate environments.

**Comment by ahmed-abdelrazek** [comment]
@paulius-valiunas you can always wait for windows 12 when this one flops and they rush to release the next one
but you can try https://github.com/builtbybel/ThisIsWin11 and https://www.stardock.com/products/start11/

**Comment by Biswa96** [comment]
This question may seem to be off-topic but it may hint something with WSLg. Can anyone running **Windows 10 non-insider stable build** check if the `init` binary contains `boot.command` string in it with this command `grep -ai boot.command /init`? If yes please reply with the full Windows OS build number.

**Comment by ahmed-abdelrazek** [comment]
@Biswa96 returns nothing 
![image](https://user-images.githubusercontent.com/7858946/141311689-304f5c29-c771-496f-9c8f-2f02d362bb5f.png)

![image](https://user-images.githubusercontent.com/7858946/141311511-21ccc0a9-f36f-4139-bb30-6834e81cb147.png)

**Comment by thany** [comment]
> 
> 
> I might be able to upgrade, but the problem is I don't want to. Windows 11 has horrible UX (excessive padding, rounded corners wasting space everywhere, lack of taskbar customization, menus that move around and block any form of muscle memory etc.) and it will take a while for all these problems to be fixed, if they're ever fixed at all. I'm still hoping for some 3rd party customizations to appear, fixing some of the glaring UX problems that Microsoft is clearly not going to fix. Isn't it ironic that I found out about WSLG by watching the VS2022 launch event where most of the demos were made by Microsoft developers using Windows 10 with customized taskbars? Microsoft, even your own developers don't use Windows 11, so how can you expect other developers to adopt it early?
> 
> Enough rant. My point is, WSLG not supporting Windows 10 is a huge problem and _not_ only in corporate environments.

Yes.

It's also just plain bullshit. There is *absolutely* no logical or technical reason that Windows 10 couldn't do this. I'm willing to bet that MS has to do way more to get it working in Windows 11, compared to Windows 10.

**Beause it already worked fine in the Insiders version. SIGH.**

So either Microsoft is using every excude in their book to force people to upgrade before they're ready (wouldn't be a first, ey, Microsoft?!) or this feature has been built by a team of brilliant experts that are now fired, and been replaced by newcomer interns.

So for the rant, but this is just absolutely fucking ridiculous. Period.

**Comment by Biswa96** [comment]
Testing WSLg (kind of) in my Windows 10 21H1 x64 system. Need to figure out many stuff.

![Capture](https://user-images.githubusercontent.com/31443074/142042636-676d4c8a-92ac-4cb3-af88-f0d293b30298.PNG)

**Comment by spronovo** [comment]
Hey folks, please remember to keep the discussion opened and respectful. 

Our target audience for WSL/WSLg are developers. We understand that not all developers can move to Windows 11 due to enterprise policy outside of their control or even if they can upgrade, may chose not to for various other reasons. 

We've heard loud and clear that the community would like to see WSLg supported on Windows 10. From my perspective, this is a great problem to have, it means we built something that folks love enough to be passionate about it, so thank you :-).

Getting WSLg supported on Windows 10 seamlessly isn't as trivial as I've seen it claimed here and in many other channels. Let me assure you there is no evil master plan blocking it. Quite the contrary, this is something the WSL & WSLg team would like to see happened as well.

The idea that it was available in Windows 10 before Windows 11 was introduced and was explicitly blocked after the introduction of Windows 11 for some evil plan isn't accurate... a better way to think about this is that the "Windows 10" build we had been flighting for a while were really Windows Next build under development / or Windows 11 in disguised if you prefer! Windows 11 without some of the UI pieces so those could be revealed as a surprise later. As with all software, changes are incremental build to build... and the Windows Next builds we had been flighting WSLg on for a few months were really the branch/basis that would become Windows 11. The actual Windows 10 branch (e.g. 21h2) never had any of those changes in the first place, so nothing was ever removed from Windows 10.

With Windows 10 21h2 update we did a lot of work to port WSL vGPU support back to Windows 10. We did have to move a lot of code back to the actual Windows 10 branch, work with our partners to get their driver working and ready against this older version of WDDM and validate the whole thing. We did this in recognition that many of our developers wanting to use this would be running Windows 10.

For WSLg, there were additional changes that were required but impractical to get in Windows 10 through the 21h2 servicing process for various reasons. In particular changes to mstsc, the RDP client WSLg uses, which are always tricky to validate for regression as this component is use in so many scenarios. Changes to the WSL Core infrastructure to support the system distro, etc... 

You may have recently seen our blog post on [WSL coming to the store](https://devblogs.microsoft.com/commandline/a-preview-of-wsl-in-the-microsoft-store-is-now-available/). This is something we've been planning and working on for a while and are very excited about. This is going to be the new way we will be shipping WSL going forward. Once it is ready, WSL in the store will be replacing the Windows inbox/built-in WSL, but continue to provide exactly the same experience to our developers. This will give us much more flexibility in term of introducing new functionality without being tied to the Windows schedule or release cycle and supporting that functionality uniformly across Windows version (when possible). 

The version of WSLg shipped in the WSL store package is using a new RDP client (msrdc) which is distributed along with that WSL package itself. msrdc has been updated for WSLg support and this mean WSLg no longer has a dependency on the Windows 10/11 inbox mstsc client. The WSL Core infrastructure necessary to host the system distro is also part of the WSL package itself and are no longer reliant on things built into Windows! In other words, the version of WSLg shipped in the WSL store package can support both Windows 10 and Windows 11!

However, the WSL Store package isn't (yet) compatible with Windows 10. It's not just a matter of "forcing" an install... there are changes necessary into Windows 10 itself to allow the WSL Core infrastructure in the store package to seamlessly take over from the old inbox infrastructure built into Windows 10 in order to provide the same seamless experience. This is something we are exploring, but we are always very careful about making promise that require changes to Windows and that have to go through our servicing process (think Windows Update) . Given billions of devices literally run Windows 10 and get those updates, we always try to be super careful when bringing changes to servicing to ensure we don’t break the world, are very picky about what we do bring back and have to go through a lengthy approval process for good reason, we need check and balances before pushing changes that can have very big impact on our install base.

Although I can’t make any official promise, if you keep politely asking and upvoting request for WSLg in Windows 10 and are patient you may eventually get rewarded 😊.

**Comment by ahmed-abdelrazek** [comment]
honestly at this point i don't care enough i already got another PC to run Linux on 24/7 
and about the evil plans it wouldn't be the first time Microsoft did something like that just this month
but i will take that explanation and run with it good enough i guess

**Comment by exitdown** [comment]
thanks @spronovo good to see the wheels turning int he background. :)

**Comment by paulius-valiunas** [comment]
Thanks @sprnovo, that is great news! I never boughtthe whole "it was removed from Windows 10" thing, sounds like a conspiracy theory to me. But I _was_ worried that Microsoft would pull a Windows 10 again and refuse to backport any features because the upper management seems to be very pushy about getting people to use latest versions nowadays. I'm very glad to hear that the possibility is not completely blocked by some corporate policy and that someone is working on it.

**Comment by zakius** [comment]
> Windows 11 without some of the UI pieces so those could be revealed as a surprise later.

so, everything good that 11 brings without all the bad parts? I'd say it should be just shipped this way but oh well, I understand it's not your decision to make and I'm glad to see you're still fighting to get it working

**Comment by crramirez** [comment]
> WSL vGPU support back to Windows 10

@spronovo apart from the vGPU ported to Windows 10. There are other things backported to Win 10 21h2 ? I recently saw --cd. I would like to understand now beyond WSLg what is supported in Windows 10 and what in 11 I lost track of in release notes.

**Comment by a-stana** [comment]
> 
> 
> Hey folks, please remember to keep the discussion opened and respectful.
> 
> Our target audience for WSL/WSLg are developers. We understand that not all developers can move to Windows 11 due to enterprise policy outside of their control or even if they can upgrade, may chose not to for various other reasons.
> 
> We've heard loud and clear that the community would like to see WSLg supported on Windows 10. From my perspective, this is a great problem to have, it means we built something that folks love enough to be passionate about it, so thank you :-).
> 
> Getting WSLg supported on Windows 10 seamlessly isn't as trivial as I've seen it claimed here and in many other channels. Let me assure you there is no evil master plan blocking it. Quite the contrary, this is something the WSL & WSLg team would like to see happened as well.
> 
> The idea that it was available in Windows 10 before Windows 11 was introduced and was explicitly blocked after the introduction of Windows 11 for some evil plan isn't accurate... a better way to think about this is that the "Windows 10" build we had been flighting for a while were really Windows Next build under development / or Windows 11 in disguised if you prefer! Windows 11 without some of the UI pieces so those could be revealed as a surprise later. As with all software, changes are incremental build to build... and the Windows Next builds we had been flighting WSLg on for a few months were really the branch/basis that would become Windows 11. The actual Windows 10 branch (e.g. 21h2) never had any of those changes in the first place, so nothing was ever removed from Windows 10.
> 
> With Windows 10 21h2 update we did a lot of work to port WSL vGPU support back to Windows 10. We did have to move a lot of code back to the actual Windows 10 branch, work with our partners to get their driver working and ready against this older version of WDDM and validate the whole thing. We did this in recognition that many of our developers wanting to use this would be running Windows 10.
> 
> For WSLg, there were additional changes that were required but impractical to get in Windows 10 through the 21h2 servicing process for various reasons. In particular changes to mstsc, the RDP client WSLg uses, which are always tricky to validate for regression as this component is use in so many scenarios. Changes to the WSL Core infrastructure to support the system distro, etc...
> 
> You may have recently seen our blog post on [WSL coming to the store](https://devblogs.microsoft.com/commandline/a-preview-of-wsl-in-the-microsoft-store-is-now-available/). This is something we've been planning and working on for a while and are very excited about. This is going to be the new way we will be shipping WSL going forward. Once it is ready, WSL in the store will be replacing the Windows inbox/built-in WSL, but continue to provide exactly the same experience to our developers. This will give us much more flexibility in term of introducing new functionality without being tied to the Windows schedule or release cycle and supporting that functionality uniformly across Windows version (when possible).
> 
> The version of WSLg shipped in the WSL store package is using a new RDP client (msrdc) which is distributed along with that WSL package itself. msrdc has been updated for WSLg support and this mean WSLg no longer has a dependency on the Windows 10/11 inbox mstsc client. The WSL Core infrastructure necessary to host the system distro is also part of the WSL package itself and are no longer reliant on things built into Windows! In other words, the version of WSLg shipped in the WSL store package can support both Windows 10 and Windows 11!
> 
> However, the WSL Store package isn't (yet) compatible with Windows 10. It's not just a matter of "forcing" an install... there are changes necessary into Windows 10 itself to allow the WSL Core infrastructure in the store package to seamlessly take over from the old inbox infrastructure built into Windows 10 in order to provide the same seamless experience. This is something we are exploring, but we are always very careful about making promise that require changes to Windows and that have to go through our servicing process (think Windows Update) . Given billions of devices literally run Windows 10 and get those updates, we always try to be super careful when bringing changes to servicing to ensure we don’t break the world, are very picky about what we do bring back and have to go through a lengthy approval process for good reason, we need check and balances before pushing changes that can have very big impact on our install base.
> 
> Although I can’t make any official promise, if you keep politely asking and upvoting request for WSLg in Windows 10 and are patient you may eventually get rewarded 😊.

UPVOTE!

**Comment by eALduderino** [comment]
WSLg should be supported on any platform running WSL.

**Comment by seffyroff** [comment]
> Hey folks, please remember to keep the discussion opened and respectful.

Yes, please folks.

> 
> ...some evil plan isn't accurate... a better way to think about this is that the "Windows 10" build we had been flighting for a while were really Windows Next build under development ...
> 

I think this is the crux of the matter.  Me and many other people were working with a build of Windows that was ideal, did everything we wanted and was seemingly arriving in 21H2.  I hope you'll appreciate that this still reads like spin to me.  Don't backport anything. Move Windows 10 forward instead.  What Windows 11 represents is a barrier to development, not a stairway to the code fairies ready to grant our wishes.  The Android support is a damp squib, the UI changes are unwanted, and should be optional. C'mon, it's a theme really.  The TPM requirement should really be the developer's choice.  There's plenty of scenarios where TPM has little use. Windows 11 makes no sense - it's got casual consumer-centric headline features that they don't know they want, and footnote dev features that developers do know they want, but not at the expense of us already having what we'd gotten setup as an ideal work environment, and a huge improvement in Windows as a versatile development platform that was capable of challenging competitors and building a new ecosystem, which is now being effectively dismissed as not a profit center (really, if not that, then what is it?  I see no other reason that the arbitrary cutoff would happen at this high point to stop us in our tracks).  

> Although I can’t make any official promise, if you keep politely asking and upvoting request for WSLg in Windows 10 and are patient you may eventually get rewarded 😊.

This could have been a much more trivial transition if the 'hidden' features to 'surprise' us were more than an unfinished widget set and an advert for Amazon/Samsung.

**Comment by paulius-valiunas** [comment]
I always thought the Insider program was meant to prevent such "surprises" by gathering feedback before releasing controversial features... guess I was wrong, this one just came out without warning

**Comment by Foosec** [comment]
Where can one get this Win11 build without the updated UI then? :)

**Comment by ahmed-abdelrazek** [comment]
@Foosec you can search for a tool that let you download windows 10 build 21390 as iso 😉 but not recommended 😓

**Comment by kmshort** [comment]
I'm using WSL2 in my datascience lab, and need wslg just for the "wsl --mount" option so I can attach a massive ext4 formatted portable data drive to my wsl2 instance!!!! plus.. many other things. :) but primarily mounting ext4.

**Comment by krutzy62** [comment]
I would like that spronovo for at least a very long and nicely worded explanation.  I for one wanted to know why more than whine.  That is more meaningful and actually makes me feel positive MS.
I really would like to politely ask most respectively to please work to back porting WSLg to windows 10 with the app.  That seems logical to me.
Thank again.

**Comment by ahmed-abdelrazek** [comment]
i agree for the most part the explanation was very good but the ending doesn't set right with me it's more like belittling with empty promise basically saying set there and be a good boy and if we feel like it we may give you the feature that we promised in the first place someday in the future maybe
if that answer was in good faith and the true official stand of Microsoft then where is my official tabbed file explorer
Microsoft put itself in this position by over promising, weak planning and rush decisions -get windows 11 out as fast as possible- but i digress

**Comment by Hashbrown777** [comment]
> I'm using WSL2 in my datascience lab, and need wslg just for the "wsl --mount" option so I can attach a massive ext4 formatted portable data drive to my wsl2 instance!!!! plus.. many other things. :) but primarily mounting ext4.

Yeah, I dont even care about the g in wsl, I'm always CLI when I dev; but I need `wsl --mount --bare`. All the reasons for it being hard to backport the graphical stuff should be a non-issue in this regard: `--mount` shouldn't be tied to the GUI stuff since it appeared in Win10 Insider before wslg was released.

**Comment by ZaDarkSide** [comment]
Just backport the damn thing from Win 11 to Win 10, I know it's possible, the update system allows it, heck it allows rewriting the whole OS, but I just hear excuses, I can't even run Win 11 even if I wanted, it complains about TPM 2.0 and...that's how much Win 11 sucks for everybody else, bottom line nobody likes Win 11 not developers, not gamers, not normal people period, but that is a whole another matter, anyway better than switching from Win 10 to Win 11 to get WSLg, better is to switch from Win 10 straight to Linux and emulate Win as needed instead of the other way around and ditch this headache. Also don't get me started on Visual Studio and C# another Microsoft product that got into the hot reload controversy https://www.youtube.com/watch?v=L9_pzDjw9HQ doing pretty much the same thing, giving something and then taking it back and putting it behind some paywall or bs roadblock.

> Although I can’t make any official promise, if you keep politely asking and upvoting request for WSLg in Windows 10 and are  patient you may eventually get rewarded 😊.

That will never happen I'm pretty sure. They killed Win 10 for no good reason, nobody cares anymore.

**Comment by Masamune3210** [comment]
Go take your ranting somewhere else, please. The update system cant "rewrite" everything, its not magic. Besides, what you are literally ragging on about the update system being able to do was already done, its called a major upgrade. Just because for some reason people have an obsession with hating 11 doesn't mean its not literally the same thing that's been happening since 8

**Comment by SunJun8** [comment]
The only thing preventing me from using win11 is the terrible explore and ui

**Comment by ZaDarkSide** [comment]
@Masamune3210 nobody knows if its a minor upgrade, major upgrade or anything in between, just because they say "it's not trivial" I really don't believe that because, the earlier versions of WSLg worked on Win 10, all we know is that WSLg will not work in Win 10 now or ever because nobody is developing stuff anymore for Win 10, is a dead operating system with only security patches planned until the end of life, also the update system upgraded you through some major rewrites of the OS if you don't remember, yes it's not black magic, but can replace all core system files with ease like has done in the past, and in the end my irrational hate for Win 11 also extends to Win Vista and Win 8, Win Me, maybe it's irrational or just maybe they were BAD releases, just look at the Win 11 adoption rate stats, no need to believe me about anything.

@SunJun8 use this https://github.com/valinet/ExplorerPatcher you will thank me later

**Comment by Masamune3210** [comment]
It didnt work on 10, it never worked or could work on 10's kernel and core systems. As stated by Microsoft devs themselves many times now, the 10 Previews that people keep harping on about were 11 in disguise, they were basically 11 without any of the new UI or apps. Of course 10 isn't getting any new major feature changes, that's how new versions work. Yes, the update system can change some most core system files, but at that point, you are basically just running a new version of Windows, it's not 10 anymore. It is very irrational hate, one that comes up literally every. single. time. a new major version of Windows comes out. The only major upgrade that hasn't had vitreous hate surrounding it is XP.  There is almost never any point to it other than new thing bad

**Comment by SunJun8** [comment]
> It didnt work on 10, it never worked or could work on 10's kernel and core systems. As stated by Microsoft devs themselves many times now, the 10 Previews that people keep harping on about were 11 in disguise, they were basically 11 without any of the new UI or apps. Of course 10 isn't getting any new major feature changes, that's how new versions work. Yes, the update system can change some most core system files, but at that point, you are basically just running a new version of Windows, it's not 10 anymore. It is very irrational hate, one that comes up literally every. single. time. a new major version of Windows comes out. The only major upgrade that hasn't had vitreous hate surrounding it is XP. There is almost never any point to it other than new thing bad


Helped me a lot, you know what? I even needed to confirm the file was deleted by pressing F5

**Comment by zakius** [comment]
@Masamune3210 whatever was it called it could be shipped with old UI or the new kernel shipped for 10, doesn't matter, just don't tie highly sought for features with UX regressions like that

though it's not a decision wslg team (the only inside people who displayed any kind of care about these issues for now) can make so... unlucky I guess

**Comment by Masamune3210** [comment]
A new kernel for 10 is exactly what 11 is. The UI is entirely subjective, if you don't like it so be it but it has nothing to do with this issue.

**Comment by paulius-valiunas** [comment]
> A new kernel for 10 is exactly what 11 is. The UI is entirely subjective, if you don't like it so be it but it has nothing to do with this issue.

@Masamune3210 excuse me but I think you're wrong here. Windows 11 is not just a new kernel, it has a huge baggage of unwanted anti-features. Insider builds have shown that the new kernel or whatever you want to call it works just fine with the old and superior UI. Your comment would only make sense if the new UI was optional and decoupled from the kernel updates - but sadly it isn't.

**Comment by Masamune3210** [comment]
Ok, its a new kernel and a bunch of bad ideas thrown in by the big wigs. And I never said the new kernel and core files are linked to the new UI. But sometimes you either have to deal with what hand you got or know when to fold

**Comment by paulius-valiunas** [comment]
> Ok, its a new kernel and a bunch of bad ideas thrown in by the big wigs. And I never said the new kernel and core files are linked to the new UI. But sometimes you either have to deal with what hand you got or know when to fold

but they _are_ "linked" in the sense that you can't get them separately. And I don't gamble, so I'll just stick with Windows 10. And maybe a Linux VM. Or the other way around. Anything but Windows 11.

**Comment by Masamune3210** [comment]
Suit Yourself

**Comment by ZaDarkSide** [comment]
Does anyone remember when Win 10 had DirectX 12 and you couldn't install it on any other older version of Windows? Exactly the same deal with WSLg.

**Comment by SunJun8** [comment]
> Ok, its a new kernel and a bunch of bad ideas thrown in by the big wigs. And I never said the new kernel and core files are linked to the new UI. But sometimes you either have to deal with what hand you got or know when to fold

Because I had to live with these "new" features for wslg

**Comment by paulius-valiunas** [comment]
That is kind of similar, but Windows 10 was an update oriented towards "gamers" in general so that made more sense. That feature also required new graphics cards, which meant people couldn't use it with their old computers anyway. And the UI updates that came with it actually added productivity instead of taking it away. The DirectX 12 update was a part of a whole, and that made sense. But Windows 11 is not developer-oriented, in fact it doesn't have any significant new features or a "theme" at all other than being Windows 10 in a new skin. And it's not like WSLg or the mount option require any new hardware, these features could work just fine on computers that don't match Windows 11 system requirements, and now they don't.

**Comment by ZaDarkSide** [comment]
> It didnt work on 10, it never worked or could work on 10's kernel and core systems. As stated by Microsoft devs themselves ?>  many times now, the 10 Previews that people keep harping on about were 11 in disguise, they were basically 11 without any of the new UI or apps. 

@Masamune3210 i'm pretty sure the Win 10 Previews were for upcoming patches for Win 10 that were cancelled when they pivoted to make Win 11. So I still call bs on your narrative.

**Comment by ZaDarkSide** [comment]
> That is kind of similar, but Windows 10 was an update oriented towards "gamers" in general so that made more sense. That feature also required new graphics cards, which meant people couldn't use it with their old computers anyway. And the UI updates that came with it actually added productivity instead of taking it away. The DirectX 12 update was a part of a whole, and that made sense. But Windows 11 is not developer-oriented, in fact it doesn't have any significant new features or a "theme" at all other than being Windows 10 in a new skin. And it's not like WSLg or the mount option require any new hardware, these features could work just fine on computers that don't match Windows 11 system requirements, and now they don't.

@paulius-valiunas Win 11 is oriented towards "gamers" also only this time it does less FPS in any benchmark ROLF

**Comment by Masamune3210** [comment]
1. That's just wrong lol, older UI is used all the time on test builds because getting things working is more important than testing reflection effects or whatever
2. Literally Microsoft themselves have come out and said that 10 Previews were basically 11 without the new UX or apps
3. Go back to the 90's Rolf

**Comment by ZaDarkSide** [comment]
> 1. That's just wrong lol, older UI is used all the time on test builds because getting things working is more important than testing reflection effects or whatever
Keep telling yourself that.
> 2. Literally Microsoft themselves have come out and said that 10 Previews were basically 11 without the new UX or apps
By that logic Windows 11 Previews are actually Windows 12 without new UX
> 3. Go back to the 90's Rolf
Software at least was not kept from users for stupid reasons.

**Comment by Hashbrown777** [comment]
> it's not like WSLg or the mount option require any new hardware, these features could work just fine on computers that don't match Windows 11 system requirements

The mount option literally does work on Win10, but only on insider previews. Disregarding wslG, the next full (non-insider) release of wsl was made win11-only. I'm using it right now. But since they've coupled the GUI stuff with the disk mount release (I doubt the code itself is coupled in any way whatsoever) there's no way to get the mount functionality on a standard version of win10.

**Comment by Masamune3210** [comment]
I'm not sure why I even bothered trying....

**Comment by Foosec** [comment]
Gotta say im with the masses on this one, there are some really nice features added to Win11, but bundeled with the horrible UI decisions that make me click more and do less, i guess ill just do what i did with vista and wind8 and just wait for win12 to roll around :) Hopefully with the UI regressions well gone

**Comment by ZaDarkSide** [comment]
More proof that it should work on Windows 10 https://developer.nvidia.com/cuda/wsl/download

![image](https://user-images.githubusercontent.com/291278/149756968-4cab1d9d-4f09-4ac7-812d-ded77756c3c9.png)

**Comment by paulius-valiunas** [comment]
> More proof that it should work on Windows 10 https://developer.nvidia.com/cuda/wsl/download
> 
> ![image](https://user-images.githubusercontent.com/291278/149756968-4cab1d9d-4f09-4ac7-812d-ded77756c3c9.png)
> 
> These drivers work on Windows 10 AND 11

keep in mind it requires an Insider build, which, as some people here have argued, can be considered an early Windows 11 build. I think it was already well established that WSLg works on Insider builds, so this brings nothing new to the discussion. As for whether we can consider those builds "Windows 10" or "Windows 11", I think we can all agree to disagree. IMHO that's besides the point anyway, given current implementation of "Feature Updates" it just doesn't make a difference.

**Comment by ZaDarkSide** [comment]
> keep in mind it requires an Insider build, which, as some people here have argued, can be considered an early Windows 11 build. I think it was already well established that WSLg works on Insider builds, so this brings nothing new to the discussion. As for whether we can consider those builds "Windows 10" or "Windows 11", I think we can all agree to disagree. IMHO that's besides the point anyway, given current implementation of "Feature Updates" it just doesn't make a difference.

YOU MISSED THE MOST IMPORTANT PART!

It says that the driver is also available to Microsoft Insider Members, obviously on that page it doesn't checks if you are one or not and urges you to use Windows Update but surprise, surprise the installer that you download from there installs fine on Windows 10 Retail aka non Preview and non Insider builds.

**Comment by paulius-valiunas** [comment]
> > keep in mind it requires an Insider build, which, as some people here have argued, can be considered an early Windows 11 build. I think it was already well established that WSLg works on Insider builds, so this brings nothing new to the discussion. As for whether we can consider those builds "Windows 10" or "Windows 11", I think we can all agree to disagree. IMHO that's besides the point anyway, given current implementation of "Feature Updates" it just doesn't make a difference.
> 
> YOU MISSED THE MOST IMPORTANT PART!
> 
> It says that the driver is also available to Microsoft Insider Members, obviously on that page it doesn't checks if you are one or not and urges you to use Windows Update but surprise, surprise the installer that you download from there installs fine on Windows 10 Retail aka non Preview and non Insider builds.

Well, if you think that's "the most important part", you should have at least briefly mentioned it in your comment.

But does that mean all the changes needed for WSLg are there? I don't think so. It doesn't mean everything WSLg needs is there. Here's an explanation from the beginning of this discussion:

> With Windows 10 21h2 update we did a lot of work **to port WSL vGPU support back to Windows 10**. We did have to move a lot of code back to the actual Windows 10 branch, work with our partners to get their driver working and ready against this older version of WDDM and validate the whole thing. We did this in recognition that many of our developers wanting to use this would be running Windows 10.
> 
> For WSLg, **there were additional changes that were required** but impractical to get in Windows 10 through the 21h2 servicing process for various reasons. In particular changes to mstsc, the RDP client WSLg uses, which are always tricky to validate for regression as this component is use in so many scenarios. Changes to the WSL Core infrastructure to support the system distro, etc...

Full comment here: https://github.com/microsoft/wslg/issues/347#issuecomment-971126353

**Comment by ZaDarkSide** [comment]
> But does that mean all the changes needed for WSLg are there? I don't think so. It doesn't mean everything WSLg needs is there. Here's an explanation from the beginning of this discussion:

Well the only bottleneck I keep seeing from getting WSLg to run on Win 10 is Microsoft not willing to backport some code from Win 11, the 3rd-party vendors already provide the needed drivers that already work on Win 10 Retail.

Anyway I wouldn't worry too much, like always there will be some smart guy that is gonna fork the repo and hack his way around the missing parts to make something at least workable as always to keep us happy 😄

**Comment by paulius-valiunas** [comment]
> > But does that mean all the changes needed for WSLg are there? I don't think so. It doesn't mean everything WSLg needs is there. Here's an explanation from the beginning of this discussion:
> 
> Well the only bottleneck I keep seeing from getting WSLg to run on Win 10 is Microsoft not willing to backport some code from Win 11, the 3rd-party vendors already provide the needed drivers that already work on Win 10 Retail.
> 
> Anyway I wouldn't worry too much, like always there will be some smart guy that is gonna fork the repo and hack his way around the missing parts to make something at least workable as always to keep us happy 😄

No, they don't provide anything. If you had read the comment I linked, or at least the highlighted phrases in my quote, you'd understand that vGPU support was backported by Microsoft themselves, and that's what Nvidia is using for their CUDA stuff. You'd also know that vGPU support alone is not the only requirement for WSLg.

**Comment by ZaDarkSide** [comment]
> No, they don't provide anything. If you had read the comment I linked, or at least the highlighted phrases in my quote, you'd understand that vGPU support was backported by Microsoft themselves, and that's what Nvidia is using for their CUDA stuff. You'd also know that vGPU support alone is not the only requirement for WSLg.

How do you know? As far as I know the driver vendors do need to provide or enable in their drivers the GPU Passthrough to be possible so you can pass down the GPU to a Virtual Machine.

https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/plan/plan-for-gpu-acceleration-in-windows-server

![image](https://user-images.githubusercontent.com/291278/149765500-08f478c8-c0c8-4302-8854-0a8b4b5fcdbd.png)

Also vGPU was part of Remote Desktop called formerly RemoteFX vGPU which was a passthrough driver for VMs written by Microsoft and also cancelled cause of "security concerns", a disabled by default flag was not enough, they had to completely remove the feature.

https://docs.microsoft.com/en-us/windows-server/virtualization/hyper-v/plan/plan-for-gpu-acceleration-in-windows-server

**Comment by paulius-valiunas** [comment]
This screenshot you sent is about running a Windows VM on a Linux host. Completely unrelated to WSL.

And I don't think he was referring to RemoteFX vGPU. Again, that is something for Windows Server running inside a VM. Unrelated to our case.

**Comment by ZaDarkSide** [comment]
> This screenshot you sent is about running a Windows VM on a Linux host. Completely unrelated to WSL.
> 
> And I don't think he was referring to RemoteFX vGPU. Again, that is something for Windows Server running inside a VM. Unrelated to our case.

Yes totally unrelated, it's like WSL is not based on a thin Hyper-V and Hyper-V isn't an evolution of Virtual PC at all, and all these technologies I listed are unrelated with each other at all and RemoteFX vGPU didn't had any relation to Hyper-V and so on, they are completely different fields of computer science. 😄

**Comment by paulius-valiunas** [comment]
@ZaDarkSide let me put it very simply: you're talking about technologies that are relevant for Windows inside Linux, not Linux inside Windows. It's like saying a graphics card can capture video input just because it has an HDMI port. Sure, it's definitely the same field of computer science, but it's just the other way around of what you want to do.

**Comment by ZaDarkSide** [comment]
> @ZaDarkSide let me put it very simply: you're talking about technologies that are relevant for Windows inside Linux, not Linux inside Windows. It's like saying a graphics card can capture video input just because it has an HDMI port. Sure, it's definitely the same field of computer science, but it's just the opposite of what you want to do.

no comment, you just took that one thing that wasn't about linux in windows and you just got stuck there

**Comment by paulius-valiunas** [comment]
> > @ZaDarkSide let me put it very simply: you're talking about technologies that are relevant for Windows inside Linux, not Linux inside Windows. It's like saying a graphics card can capture video input just because it has an HDMI port. Sure, it's definitely the same field of computer science, but it's just the opposite of what you want to do.
> 
> no comment, you just took that one thing that wasn't about linux in windows and you just got stuck there

first you say I missed "the most important part" of your comment which wasn't even there, now you say I "took out one thing" that's actually the entirety of your comments. I have a feeling you forgot to type something again. Because I simply don't see anything relevant in your last 2 comments.

**Comment by ahmed-abdelrazek** [comment]
that got out of hand quickly but let me say this again
i don't care if it's black magic if microsoft said it's coming to windows 10 it should deliver
it's not the end user fault what they do in their development process and i would of understood if they canceled the feature but the feature is here in windows 11 -they won't even say it's not coming to windows 10 directly just vague wording- and some people here treat windows like it's that small open source and free project and not paid os

as i pointed out before windows 10 21h2 comes with gpu comute for wsl which needed the same os builds as wslg 
![image](https://user-images.githubusercontent.com/7858946/149794166-52a93ce6-a74e-43ad-80a8-9ebc41855e61.png)
![image](https://user-images.githubusercontent.com/7858946/149794220-6e2a445b-febc-437e-a8f7-7aae9bbc06c5.png)

microsoft has a history of pulling that stuff if they can do all of this 
https://www.pcgamer.com/microsoft-finally-ports-directx-12-to-windows-7-but-only-for-certain-games/
https://devblogs.microsoft.com/directx/porting-directx-12-games-to-windows-7/
https://www.makeuseof.com/directstorage-windows-10-get-your-pc-ready/
https://devblogs.microsoft.com/directx/directstorage-developer-preview-now-available/
they can do this too even if it was limited

stop being so apologetic for a corporation -they can talk for themselves- just imagine if we were asking for 'windows subsystem for android' then i would of understand the excuses

**Comment by ahmed-abdelrazek** [comment]
the whole windows team needs a break

**Comment by docteurzoidberg** [comment]
please don't forget the developers stuck on win10 because of their old core i7 cpu

**Comment by kmshort** [comment]
I don't understand the hate with Win11 tbh. Yeah the hoops you have to go through when right clicking something are a PITA, but everything else is mostly fine. Get over it. I think @craigloewen-msft has been pretty up front with how the wslg development has coincided with Win10/11 insider previews and the evolving dependencies that come along with that mean that wslg will break on a mismatched kernel. Maintaining two wslg versions for 2 different versions of windows requires more work and more people. I don't know what @craigloewen-msft has for a salary budget because it probably isn't a money spinner for msft.

Anyway, the windowing side of wslg can be done using Xterm emulators in Wsl2. I still do that on Win10 and use things like R studio. Heck, I've used wsl2 to ssh into a remove server, and forward graphical xterm windows from the remote server to wsl2, to my desktop. It's slow AF but it works.

What I simply want is wsl --mount capability so I can mount separate Ext4 HDDs and network locations, and of course (and it's not in wslg yet) systemd.

**Comment by ahmed-abdelrazek** [comment]
@kmshort yeah it must be nice when you live in country where all Microsoft service are available to you and you don't need anything other than Microsoft forced on you have baked products because you don't need anything else 
you literally have to go extra step from vista to see your widgets which they removed from 8 and 10 because no body was using them but now that app did it we must have them as well
it must be nice to have functioning store 
![image](https://user-images.githubusercontent.com/7858946/149876943-baf79317-e057-4763-9af2-b17fe8584804.png)

it must be nice to not have to use customized srt or ass subtitles or other basic media player function which existed in windows media player but not movies and tv app and it just works for you

it must be nice to like chromium based browsers and like edge and not feel it's forced down your throat or something

it must be nice to just use what ever default crap Microsoft gives you without the need to customized you workflow or anything  because they know better so no need to do something like this
![image](https://user-images.githubusercontent.com/7858946/149877632-165a37ac-e412-45a2-b60c-afecdb0bca0f.png)

it must be nice not having any muscle memory doing the same stuff for more than 20 year which actually just works for Microsoft to just say f* it we are changing all of that just because android and ios lol

it must be nice living in a world where everybody only uses surfaces with intel 13th gen cpus and have unlimited gigabit internet connection or 5g in the middle of global pandemic price hikes and chip shortages 

and i can still go on but you don't have to understand you are not Microsoft all you need to know is people hate it for different reasons and they entitled to their opinions if you use it and like it good for you but don't tell others ah "it's actually good you are wrong"
i finished cyberpunk 2077 three times once for each path in the game played it after a week of release had issues with the game only three times which is basically nothing for 300+ hour of play time i don't go around telling people that they it's bad you are wrong becuase i like it and i don't see those issues don't be that Karen
but enough about windows 11

**Comment by Foosec** [comment]
Everyone also seems to miss out on the fact that the whole arbitrary restrictions on CPUs thing, is probably just a deal with OEMs to induce a hardware cycle. But here is the kicker, from an envoirmental standpoint, that is absolutely horrible, my i7 3770k still runs all my games fine, someones 6th or 7th gen cpu is more than fast enough, especially if its an office PC. We already have a silicon shortage, lets not make it worse by forcing a hardware cycle please

**Comment by ZaDarkSide** [comment]
Drop TPM 2.0 and CPU restrictions and I will gladly install Windows 11 to use WSLg and stop complaining here, it's that simple.

**Comment by ahmed-abdelrazek** [comment]
@ZaDarkSide i basically got myself 200$ 4th gen i7 desktop and made it into Linux server no windows 11 or wslg required 
also you can bypass windows 11 requirements easily there is ton of articles out there about it if you don't mind using windows 11 you can do that too

**Comment by ZaDarkSide** [comment]
> @ZaDarkSide i basically got myself 200$ 4th gen i7 desktop and made it into Linux server no windows 11 or wslg required also you can bypass windows 11 requirements easily there is ton of articles out there about it if you don't mind using windows 11 you can do that too

You can bypass them if you want your updates to one day just stop working aka no official support if you get any problems while doing that. Anyway I see Microsoft is perfectly happy to herd people away from their OS straight to Linux 😄

**Comment by ahmed-abdelrazek** [comment]
well i don't say it's a good option just another one if you really want wslg and can't or won't get another pc just for linux
as for people going away from windows i can't imagine what would happen if the steam deck is a huge success

**Comment by yume-chan** [comment]
Instead of keep complaining or arguing, why not try it yourself?

![image](https://user-images.githubusercontent.com/1330321/150192129-a69a8315-7c67-45d0-8810-2245dfef12b7.png)

Mouse input doesn't work yet, and GPU acceleration of course won't work, but it's not that hard to get something done.

Quick summary:

1. Extract WSLg system distro (`system.vhd`) from Store package on Win 11 (I believe you can also download Store package using some website from Win10, or build yourself following the official [guide](https://github.com/microsoft/wslg/blob/18884f996f5e851af4842010b434db17d3caa7aa/CONTRIBUTING.md))
2. Convert it to a `tar` archive (I use `wsl --mount` to mount the vhd, then `chroot /mnt/wsl/system /bin/tar --acls --xattrs -cpz / > ~/wslg.gz.tar` the whole folder into a gzipped `tar` then extracted it (don't want to find out how to create a uncompressed tar))
3. Install it as normal user distro on Win 10 (`wsl --import`)
5. Run `HCSDIAG list` on Windows to get WSL VM ID (the GUID)
6. Set VM ID to `WSL2_VM_ID` environment variable in WSLg distro
2. "Borrow" the `msrdc.exe` (with its friends) and `wslg.rdp` from Store package
3. Set folder path of `msrdc.exe` to `WSL2_INSTALL_PATH` environment variable in WSLg distro
4. Manually hook up folders on both WSLg distro and a normal Ubuntu distro (link `/mnt/wsl/wslg` to `/mnt/wslg`, create folders as `WSLGd` requires, link `/mnt/wslg/.X11-unix` to `/tmp/.X11-unix`)
5. Start `WSLGd`
6. Set `DISPLAY=:0` in Ubuntu distro
7. Start GUI Apps

**Comment by crramirez** [comment]
@yume-chan Store package can be downloaded from here: https://github.com/microsoft/WSL/releases

**Comment by ZaDarkSide** [comment]
> Thanks @spronovo, that is great news! I never boughtthe whole "it was removed from Windows 10" thing, sounds like a conspiracy theory to me. But I _was_ worried that Microsoft would pull a Windows 10 again and refuse to backport any features because the upper management seems to be very pushy about getting people to use latest versions nowadays. I'm very glad to hear that the possibility is not completely blocked by some corporate policy and that someone is working on it.

@paulius-valiunas oh no is that WSLg running in Win 10 in the screenshot that yume-chan posted? oh no a conspiracy...

**Comment by paulius-valiunas** [comment]
@ZaDarkSide I'd appreciate if you toned down the toxicity a bit. Let me remind you this is not 4chan.
Sure, there's a workaround. A complicated and untested workaround. I choose to believe they're working on making it official (i.e. tested and approved), and you can continue insisting there's a conspiracy and they'll never allow it to happen, if you wish so.

**Comment by docteurzoidberg** [comment]
@yume-chan wow that's super handy. i would better go straight formatting my computer than doing this

**Comment by ZaDarkSide** [comment]
> @ZaDarkSide I'd appreciate if you toned down the toxicity a bit. Let me remind you this is not 4chan. Sure, there's a workaround. A complicated and untested workaround. I choose to believe they're working on making it official (i.e. tested and approved), and you can continue insisting there's a conspiracy and they'll never allow it to happen, if you wish so.

Those were your words "conspiracy" not mine, I just quoted you, anyway for over like 20 posts you were trying to say WSLg cannot work on Win 10 there is no proof etc. it's funny how you are proven wrong at literally everything you write.
Nobody is being toxic, you just spout non-sense about a subject you know very little about, my advice is if you can't contribute with facts, keep your "conspiracy" for yourself.

**Comment by Hashbrown777** [comment]
@yume-chan Note that people wont have `--mount` because that's only in the Win11 version..
Also, `-z` is doing the gZip'ing, and the naming convention would be `.tar.gz` as it's a tarball 'wrapped' in a gzip.

I also saw no explanation on how to extract the package, but I figured it out with some research.
Here's a way to do their first 2 steps with a standard Win10:

1. Download the SDK from https://developer.microsoft.com/en-us/windows/downloads/windows-sdk/
2. Install "Windows App Certification Kit"
3. Download the WSL release from https://github.com/microsoft/WSL/releases
4. Unbundle the release with powershell using `&'C:\Program Files (x86)\Windows Kits\10\App Certification Kit\makeappx.exe' unbundle /p Microsoft.WSL_0.51.0.0_x64_ARM64.msixbundle /d <#directory#>`
5. Unpack the contained install with `&'C:\Program Files (x86)\Windows Kits\10\App Certification Kit\makeappx.exe' unpack /p Microsoft.WSL_0.51.0.0_x64.msix /d <#directory#>`
6. In WSL install libguestfs-tools (eg `sudo apt-get install libguestfs-tools`)
7. Make a directory and mount the vhd from the WSL package `mkdir wsl && guestmount -a system.vhd -i wsl`
8. Tarball it `tar --acls --xattrs --transform 's#^wsl/##' -cpf wslg.tar wsl`
9. (cleanup `umount wsl && rm -r wsl`)

Now you have your tarball you can copy back out into windows space for `--import` and follow the rest of their steps.

*But* I came here for `--mount`, not the GUI, and I'm assuming for this linux image to work your wsl EXECUTABLE will need to support it (and thus already have `--mount` in the first place) so I'm looking into maybe copying the extracted package from step 5 here directly over my wsl install and seeing if that 'just works'...

**Comment by yume-chan** [comment]
@Hashbrown777 Yeah, I didn't say this is the best way to do that, because it was 3AM I just want to do it quick (and dirty).

But you don't need any special software to extract `system.vhd` from the downloaded package. `msixbundle` and `msix` files are just zip archives:

```sh
$ file Microsoft.WSL_0.51.0.0_x64_ARM64.msixbundle
Microsoft.WSL_0.51.0.0_x64_ARM64.msixbundle: Zip archive data, at least v4.5 to extract
$ file Microsoft.WSL_0.51.0.0_x64.msix
Microsoft.WSL_0.51.0.0_x64.msix: Zip archive data, at least v4.5 to extract
```

I assumed every one knows that, maybe it's not that obvious.

---

As I mentioned before, the best way to get the `wslg.tar` is building it from source! This repository contains all source code and scripts to do it, and it's all licensed under MIT so won't cause any trouble unlike the pre-packaged one.

In [https://github.com/microsoft/wslg/blob/18884f996f5e851af4842010b434db17d3caa7aa/CONTRIBUTING.md#build-instructions](https://github.com/microsoft/wslg/blob/18884f996f5e851af4842010b434db17d3caa7aa/CONTRIBUTING.md#build-instructions) step 2.2, you can see it actually built a Docker image, and exported to a tar file. Then in 2.3 it converts the tar into vhd. So just stop at 2.2

**Comment by Hashbrown777** [comment]
Yeah I figured the zip part out and had added it as an edit. I would've assumed the gui stuff would also communicate with some executables outside the linux environment, but I guess that's what msrdc and mstsc are for and wsl itself isn't actually?

@yume-chan I guess the only way to test if your solution works is for you to show gui is working *whilst simultaneously having --mount be missing* on your system. What happens when you run `wsl --mount`, does it say "Invalid command line option: --mount"?

Instead of blindly overwriting the system32 wsl exe's with the package I tried editing the appxmanifest's min version requirement, but I couldn't get it repackaged in a way to install it. I bet if I did just overwrite them it'd work, there doesn't seem to be anything else in the package. Even if it does require some [windows] kernel stuff that isn't in this build worst case it'll blue-screen when I attempt to run wsl, but I'm hesitating on pulling the trigger just yet.

**Comment by yume-chan** [comment]
> does it say "Invalid command line option: --mount"?

Yes, of course it doesn't have `wsl --mount`

![image](https://user-images.githubusercontent.com/1330321/150531557-d28eec59-87d1-40f7-bd75-7c448b11976f.png)

> Instead of blindly overwriting the system32

Actually, when Store version of wsl was installed on Windows 11, it doesn't modify any system files (it doesn't replace `wsl.exe` in `C:\Windows\System32`), instead, `C:\Windows\System32\wsl.exe` will launch the `wsl.exe` in Store package, pass all arguments and pipe the output.

> I tried editing the appxmanifest's min version requirement

I also tried that first.

> but I couldn't get it repackaged in a way to install it

Delete `AppxMetadata` folder, `AppxBlockMap.xml` and `AppxSignature.p7x`, then run `Add-AppxPackage --Register AppxManifest.xml`. It will install on Windows 10 20H2

> there doesn't seem to be anything else in the package

No, `wslservice.exe` is a Windows Service that's very important.

> Even if it does require some [windows] kernel stuff that isn't in this build worst case it'll blue-screen

Yes, `wslservice.exe` crashes on startup on Windows 10 21H2. I bet it requires some kernel changes.

**Comment by Hashbrown777** [comment]
> Yes, `wslservice.exe` crashes on startup on Windows 10 21H2. I bet it requires some kernel changes

Bugger, well thanks for the info!
This's been a learning experience at least.

I'll just wait for some sort of action on microsoft/WSL#7940 then.

**Comment by yume-chan** [comment]
> I would've assumed the gui stuff would also communicate with some executables outside the linux environment

Forgot to mention, MS said the architecture will work on any physical or virtual Linux machines (it's transportation is standard RDP protocol, as hinted by the usage of msrdc/mstsc). WSLg will work better because it runs both Linux and Windows on the same machine and Hyper-V supports memory sharing so data doesn't need to be passed over network (and msrdc/mstsc on Windows can be started directly from Linux).

So getting WSLg to run on Windows 10 doesn't require too much kernel changes (except GPU virtualization, I don't know if all features required by WSLg was backported as mentioned before by other guys)

**Comment by MrLinDowsMac** [comment]
> Straight up, not everyone has support for Win11, E.G I don't have an TPM module and my CPU `fx 8350` is simply not supported. I simply do not see a reason why i would ever want to have a TPM module implemented. the last time i've had ransomware was like... Never, or a virus, also something i haven't had in like 8 years. Only kids who crack random ass software get this shit or people who don't know what they're doing. but quite literally a useless module for technically served people.
> 
> Also, i cannot afford to spend 800 bucks for a new freaking computer, especially in a chip shortage we have today. literally 3 times the price it used to be, even second hand.


I do have TPM 2.0 and Secure boot but still unable to update my 3-4 years old PC just because my processor intel processor is 7th gen. Infuriating. 
PCs aren't cheap in my country either, this is almost like a planned obsolescence.

**Comment by KoenT-SS** [comment]
@spronovo Is there an update on this?
I am currently running WSL2 in Windows 10 Pro 21H2 (10.0.19044 Build 19044), which works nicely (thanks for doing this BTW).
But for projects I'm working on I also need audio recording and playback, which apparently requires WSLg, and we can currently only get WSLg on Windows 11.
And it turns out I can't upgrade my machine to Windows 11 for these reasons:
- This PC must support Secure Boot
- TPM 2.0 must be supported and enabled on this PC
- The processor isn't currently supported for Windows 11

This is a very capable / powerful MacBook Pro machine though: Intel Core i7-4870HQ CPU @ 2.50 GHz 4 cores / 8 logical processors, 16 GB RAM.
I still hope to see either lower requirements for upgrading to Windows 11, or WSLg support for Windows 10 Pro users.
Thanks!

**Comment by Biswa96** [comment]
It is somehow possible to use WSLg feature in Windows 10 using this idea https://github.com/Biswa96/initrdg. But the project is not in so convenient yet for all users. If some users can provide some feedback I can try to make it usable.

**Comment by KoenT-SS** [comment]
@Biswa96 Thanks for the offer and you efforts to document this, but I tend to stay away from experimental workarounds. Still hoping for a solution supported by Microsoft.

**Comment by KamranShahid** [comment]
Wslg support in windows 10 is the need of the time. $microsoft should please consider it

**Comment by einarpersson** [comment]
@spronovo

> "We've heard loud and clear that the community would like to see WSLg supported on Windows 10. "

> "Although I can’t make any official promise, if you keep politely asking and upvoting request for WSLg in Windows 10 and are patient you may eventually get rewarded"

I have been checking in this issue now and then for the past half year. What is the status currently? Has it moved forward at all, or is this issue abandoned?

**Comment by ahmed-abdelrazek** [comment]
well they finally adding tab support for file explorer maybe all hope is not lost after all

**Comment by lornova** [comment]
I was very happy to work with Windows 10 and WSL. Although my workstation is perfectly capable of running Windows 11, I choose not to, as its user experience is unbearably terrible (no taskbar, no start menu, ugly look and feel etc.), so I guess that if full WSL support, including WSLg, will not be backported to Windows 10, I'll simply switch my main OS to Linux...

**Comment by zakius** [comment]
it's (at the time of writing this comment, may change with any "update" of w11) easier to make w11 usable (basically Explorer Patcher + 7+TT as always + few AHK scripts as always) than any Linux distro known to me, even though KDE itself is quite capable there are other issues that may be solvable, but that's a huge time investment so I wouldn't attempt that on my main work machine

**Comment by Mukul1127** [comment]
Any Updates @spronovo ?

**Comment by MrLinDowsMac** [comment]
Yes @spronovo, Any updates?

**Comment by cormacltighe** [comment]
Going to leave a comment here so I can see it when this (hopefully) gets updated

**Comment by Masamune3210** [comment]
If you want all notifications of this issue report, the subscribe button is what you want, not commenting

**Comment by thiagofranca85** [comment]
Until Microsoft doesn't re-enable WSLg natively on Windows 10 (I saw some videos and articles of people running GUI APPs during my searching, with Windows 10, only doing "wsl --update" and that's it, unfortunately, this command doesn't work for me to run WSLg after updated). I found something here that can help some people. I'm just starting learning how to program and spent a few days trying to run WSLg natively on Windows 10 (21H2 - 19044.1889) but without success (A lot of gtk-warnings and display errors, I'm new at Linux too). So I found an Application on MS Store that maybe can help some people while MS makes WSLg usable in Windows 10 again. The application name is GWSL and runs as an xServer, very simple to use for new users like me (Just install it and click on GWSL on system tray > GWSL Distro Tools then Auto-Export Display/Audio, restart Ubuntu or any other Distro installed and that's it), this can be an option atm. I hope MS enables again WSLg for Windows 10 users soon. I tried to run Windows 11 here but I don´t have driver support for my machine at the moment, then I had to go come back to Windows 10.

**Comment by cbsplinter** [comment]
Please enable this feature

**Comment by crramirez** [comment]
@thiagofranca85 telling that WSLg was disabled in Windows 10 is a source of confusion. This implies that at some point, Microsoft disabled WSLg in Windows 10, which is not true.

WSLg was introduced in Windows Insider version [21364](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-21364); note the considerable version difference between 19044 and 21364. So, everything since 19555 is part of Windows 11, not Windows 10. So, WSLg doesn't even exist in Windows 10 and needs to be BACKPORTED with all the work and risk this implies.

Many things were backported to Windows 10, like the GPU support or improvements in the wsl command.

I make this clarification to state that having WSLg in Windows 10 is not as simple as re-enable it.

Hope it helps.

**Comment by Biswa96** [comment]
> WSLg was introduced in Windows Insider version [21364](https://docs.microsoft.com/en-us/windows/wsl/release-notes#build-21364)

So, what was it called back then Windows 10 or Windows 11? I guess it was not Windows 11 at that time, right?

**Comment by crramirez** [comment]
Yes, before Windows 11 announce it was called Windows 10 Insider Version

**Comment by KamranShahid** [comment]
> Yes, before Windows 11 announce it was called Windows 10 Insider Version

All in all customers simply want that feature without windows 11. MS$ have to think about it

**Comment by thiagofranca85** [comment]
@crramirez Thanks for your reply. For users like me without the Windows Insider enabled, this feature is not available. What I said was a big misunderstood, sorry about that. Thanks for clarifying my mistakes in your reply. I´m new in Dev, Github, etc just learning a lot here too.

**Comment by zakius** [comment]
> > Yes, before Windows 11 announce it was called Windows 10 Insider Version
> 
> All in all customers simply want that feature without windows 11. MS$ have to think about it

or win11 with the 10 UI as it used to be tested before

**Comment by vathanac** [comment]
Is Wslg really necessary? Why not just dualboot a linux distro alongside windows 10?

**Comment by paulius-valiunas** [comment]
> Is Wslg really necessary? Why not just dualboot a linux distro alongside windows 10?

I hope you're joking. You can't compare running a Linux application right in the OS you're using, to dropping everything you were doing, completely shutting down your computer and booting into a completely different operating system just to run that one application. Not to mention maintaining both OS's with their updates etc.

**Comment by kmshort** [comment]
> > Is Wslg really necessary? Why not just dualboot a linux distro alongside windows 10?
> 
> I hope you're joking. You can't compare running a Linux application right in the OS you're using, to dropping everything you were doing, completely shutting down your computer and booting into a completely different operating system just to run that one application. Not to mention maintaining both OS's with their updates etc.

Now that's a mic drop. Perfectly well said.
Wslg supplements Windows with Linux. We're here to use Linux, not run Linux.

**Comment by bersbersbers** [comment]
Looks like this is fixed, see https://github.com/microsoft/wslg/issues/567#issuecomment-1319599234. I can run `xeyes` on Windows 10 21H2 now (19044.2311, to be exact).

**Comment by vathanac** [comment]
> 

I didn't said wsl isn't necessary. What I want to ask is running gui linux application on wsl really that necessary.

**Comment by paulius-valiunas** [comment]
> I didn't said wsl isn't necessary. What I want to ask is running gui linux application on wsl really that necessary.

you can now run GUI apps in WSL on Windows 10. Case closed.

**Comment by ahmed-abdelrazek** [comment]
> > I didn't said wsl isn't necessary. What I want to ask is running gui linux application on wsl really that necessary.
> 
> you can now run GUI apps in WSL on Windows 10. Case closed.

would be great if true going to try this soon

**Comment by crramirez** [comment]
> > > I didn't said wsl isn't necessary. What I want to ask is running gui linux application on wsl really that necessary.
> > 
> > 
> > you can now run GUI apps in WSL on Windows 10. Case closed.
> 
> would be great if true going to try this soon

It is true, confirmed. As long you have Windows 10 22h2

**Comment by spronovo** [comment]
Hey folks, it is indeed official and WSLg is now available for Windows 10. Please make sure that your Windows 10 installation is fully up to date and then update to the latest version of WSL in the Microsoft Store.

For more details, please see our official blog on this.

https://devblogs.microsoft.com/commandline/the-windows-subsystem-for-linux-in-the-microsoft-store-is-now-generally-available-on-windows-10-and-11/

**Comment by Brandon98SE** [comment]
### Just want to add: 
- You will need to be running [Windows 10 version 21H1, 21H2, or 22H2, or on Windows 11 21H2 with all of the November updates applied](https://devblogs.microsoft.com/commandline/the-windows-subsystem-for-linux-in-the-microsoft-store-is-now-generally-available-on-windows-10-and-11/). You will know you have this update when you check that [KB5020030](https://support.microsoft.com/en-us/topic/november-15-2022-kb5020030-os-builds-19042-2311-19043-2311-19044-2311-and-19045-2311-preview-237a9048-f853-4e29-a3a2-62efdbea95e2) is installed on Windows 10, or [KB5019157](https://support.microsoft.com/en-us/topic/november-15-2022-kb5019157-os-build-22000-1281-preview-d64fb317-3435-49ff-b2c4-d0356a51a6b0) on Windows 11.
### Current Windows LTSC WSL Version Support:
- **_NONE_** = Windows 10 Enterprise LTSB 2015 (v1507)
- **_WSL_** = Windows 10 Enterprise LTSB 2016 (v1607)
- **_WSL_** = Windows 10 Enterprise LTSC 2019 (v1809)
- **_WSL_** & **_WSL2_** & **_WSLg_** = Windows 10 Enterprise LTSC 2021 (v21H2)
